python/cudf/cudf/core/column/column.py

```diff
@@ -489,80 +491,130 @@ def __setitem__(self, key: Any, value: Any):
         """
         Set the value of self[key] to value.
 
-        If value and self are of different types,
-        value is coerced to self.dtype
+        If value and self are of different types, value is coerced to
+        self.dtype. Assumes ``self`` and ``value`` are index-aligned.
         """
 
+        # Normalize value to scalar/column
+        value_normalized: Union[cudf.core.scalar.Scalar, ColumnBase]
+        if is_scalar(value):
+            value_normalized = cudf.Scalar(value, dtype=self.dtype)
+        else:
+            value_normalized = as_column(value).astype(self.dtype)
+
+        out: Optional[ColumnBase]  # If None, no need to perform mimic inplace.
         if isinstance(key, slice):
-            key_start, key_stop, key_stride = key.indices(len(self))
-            if key_start < 0:
-                key_start = key_start + len(self)
-            if key_stop < 0:
-                key_stop = key_stop + len(self)
-            if key_start >= key_stop:
-                return self.copy()
-            if (key_stride is None or key_stride == 1) and is_scalar(value):
+            out = self._scatter_by_slice(key, value_normalized)
+        else:
+            key = as_column(key)
+            if not is_numeric_dtype(key.dtype):
+                raise ValueError("Unknown scatter map type.")
+            out = self._scatter_by_column(
+                cast("cudf.core.column.NumericalColumn", key), value_normalized
+            )
+
+        if out:
+            self._mimic_inplace(out, inplace=True)
+
+    def _scatter_by_slice(
+        self, key: Slice, value: Union[cudf.core.scalar.Scalar, ColumnBase]
+    ) -> Optional[ColumnBase]:
+        """If this function returns None, it's either a no-op (slice is empty),
+        or the inplace replacement is already performed (fill-in-place).
```

Comment:
Which code path leads to returning None because the inplace replacement is already performed? I see the explicit `return None` for an empty slice, but all the other paths seem to return columns.

Reply:
It's the `self._fill` code path, given `stride==1` and `value` is a scalar.